tests/storages_tests/test_journal.py

```diff
@@ -116,3 +128,107 @@ def test_pop_waiting_trial_multiprocess_safe() -> None:
+)
+def test_snapshot(storage_mode: str, kwargs: Dict[str, Any]) -> None:
+    with mock.patch("optuna.storages._journal.storage.SNAPSHOT_INTERVAL", 1, create=True):
+        with StorageSupplier(storage_mode, **kwargs) as storage1:
+            assert isinstance(storage1, optuna.storages.JournalStorage)
+
+            study_id = storage1.create_new_study()
+
+            # The snapshot is not saved here.
+            storage1.create_new_trial(study_id)
+
+            # The number of read logs are same thanks to `sync_with_backend` without any snapshots.
+            with StorageSupplier(storage_mode, **kwargs) as storage2:
+                assert isinstance(storage2, optuna.storages.JournalStorage)
+                assert (
+                    storage1._replay_result.log_number_read
+                    == storage2._replay_result.log_number_read
+                )
+
+            # The number of read logs are not same since `sync_with_backend` is disabled and the
+            # snapshot was not saved.
+            with mock.patch(
+                "optuna.storages._journal.storage.JournalStorage._sync_with_backend"
+            ) as m:
+                with StorageSupplier(storage_mode, **kwargs) as storage2:
+                    assert isinstance(storage2, optuna.storages.JournalStorage)
+                    m.assert_called_once()
+                    assert (
+                        storage1._replay_result.log_number_read
+                        != storage2._replay_result.log_number_read
+                    )
+
+            # The snapshot is saved here.
+            storage1.create_new_trial(study_id)
+
+            # The number of read logs are same thanks to the snapshot without `sync_with_backend`.
+            with mock.patch(
+                "optuna.storages._journal.storage.JournalStorage._sync_with_backend"
+            ) as m:
+                with StorageSupplier(storage_mode, **kwargs) as storage2:
+                    assert isinstance(storage2, optuna.storages.JournalStorage)
+                    m.assert_called_once()
+                    assert (
+                        storage1._replay_result.log_number_read
+                        == storage2._replay_result.log_number_read
+                    )
+
+            # The snapshot is saved here.
+            _ = storage1.create_new_study()
+
+            # The number of read logs are same thanks to the snapshot without `sync_with_backend`.
+            with mock.patch(
+                "optuna.storages._journal.storage.JournalStorage._sync_with_backend"
+            ) as m:
+                with StorageSupplier(storage_mode, **kwargs) as storage2:
+                    assert isinstance(storage2, optuna.storages.JournalStorage)
+                    m.assert_called_once()
+                    assert (
+                        storage1._replay_result.log_number_read
+                        == storage2._replay_result.log_number_read
+                    )
```

Comment:
It looks that this test scenario is a bit complex and depends on implementation details very much. To make it simpler and improve the maintainability, how about rewriting it like below?

https://github.com/HideakiImamura/optuna/compare/feature-journal-redis-snapshot-2...c-bata:optuna:feature-journal-redis-snapshot-2?expand=1